File: runGame.ts
import { Game } from "./game";
import { Move } from "./move";
import { Snacks } from "./snacks";
import { State } from "./state";
import { Strategy } from "./strategy";

export class RunGame {
  private snacks: Snacks;
  constructor() {
    this.snacks = new Snacks();
  }

  private oneHotMove(move: Move) {
    let maxValue = -1000;
    let maxIndex = 0;
    const result = new Move(move.data.length);
    for (let i = 0; i < move.data.length; ++i) {
      result[i] = (0.1 / (move.data.length - 1));
      if (move.data[i] > maxValue) {
        maxValue = move.data[i];
        maxIndex = i;
      }
    }
    result.data[maxIndex] = 0.9;
    return result;
  }

  // Runs the game, returns the player number who won or -1 if there is
  // no winner.
  private run(game: Game, strategies: Strategy[],
    outStates: State[], outWinProb: Float32Array[]): void {
    const states: State[] = [];

    console.assert(game.getPlayerCount() === strategies.length);
    let state = game.getInitialState();
    while (!state.isEnded()) {
      const currentPlayer = state.playerIndex;
      const move = strategies[currentPlayer].getMove(state);
      states.push(state);
      state = game.applyMove(state, move);
    }
    states.push(state);
    const winners = state.winners;
    for (const s of states) {
      outStates.push(s);
      outWinProb.push(winners);
    }
  }

  // outWinProb will be populated with indicators (0 for loss, 1 for win)
  // for the players in the game.
  collectWinData(game: Game, strategies: Strategy[],
    outStates: State[], outWinProb: Float32Array[], gameCount: number): void {
    for (let i = 0; i < gameCount; ++i) {
      this.run(game, strategies, outStates, outWinProb);
    }
  }

  getSnacks() {
    return this.snacks;
  }
}

File: py/runGame.py
import numpy as np
from numpy.lib.function_base import append
from gamePairs import GamePairs


class RunGame:
    # game play stops if either of these are achived.
    kMinWins = 100
    kMinGames = 2000

    def oneHotMove(self, move):
        maxValue = -1000
        maxIndex = 0
        result = np.zeros(len(move))
        for i in range(len(move)):
            result[i] = 0.1
            if (move[i] > maxValue):
                maxValue = move[i]
                maxIndex = i
        result[maxIndex] = 0.9
        return result

    def run(self, game, strategies, verbose=0):
        state = game.getInitialState()
        states = list()
        moves = list()
        for _ in range(game.getPlayerCount()):
            states.append(list())
            moves.append(list())
        game.startGame()
        currentPlayer = game.playerIndex
        while not game.isEnded(state):
            if verbose > 0:
                game.showState(state)
            move = strategies[currentPlayer].getMove(state)
            states[currentPlayer].append(state)
            moves[currentPlayer].append(self.oneHotMove(move))
            state = game.applyMove(state, move)
            currentPlayer = game.nextPlayer()
        winner = game.getWinner(state)
        if verbose > 0:
            game.showState(state)
        return winner, states, moves

    def collectWinData(self, game, strategies, trainingPairs):
        winCount = 0
        gameCount = 0
        # for _ in range(gameCount):
        while (winCount < self.kMinWins and gameCount < self.kMinGames):
            states = []
            moves = []
            # while len(states) < game.getPlayerCount():
            #     states.append([])
            #     moves.append([])
            winner, states, moves = self.run(game, strategies)
            for player in range(len(states)):
                isWinner = player == winner
                for i in range(len(states[player])):
                    state = states[player][i]
                    move = moves[player][i]
                    trainingPairs.append(state, move, isWinner)
            if winner == 0:
                winCount += 1
            gameCount += 1
        winRate = winCount / gameCount
        f = open("wins.txt", "a")
        f.write(str(winRate) + "\n")
        f.close()
        print("Win rate: " + str(winRate))
        return winRate
